datamade/la-metro-councilmatic
# Add reCAPTCHA to search form

## Overview

This PR adds reCAPTCHA validation to search form submissions and internal search link clicks, as there are several areas of the site that link to slices of the search data. This implementation will not expose results on direct navigation to the search page, i.e., typing `boardagends.metro.net/search` into the address bar, or clicking an external link to search results. AFAIK, this is not possible, because reCAPTCHA needs to do some work clientside to issue a token for validation serverside. The vast majority of traffic comes to the site from the homepage, so I think this should be fine.

This implementation works by adding a captcha field to the search form via [the `django-recaptcha` plug-in](https://github.com/praekelt/django-recaptcha/). The plug-in conveniently provides validation of the token for us, so we're covered serverside. The primary work was to wire up acquiring reCAPTCHA tokens and passing them with search requests on form submit and link click. I accomplished that [here](https://github.com/datamade/la-metro-councilmatic/pull/737/files#diff-61cb5d819d798e3c05ed276688d489b69eb3c559248eacb55eb1da38329fcc4f).

The flow is as follows:

- A search form is submitted, or a link to the search page is clicked.
- The corresponding event listener fires, retrieves a reCAPTCHA token, and populates a hidden input (form submission) or query parameter (link click).
https://github.com/datamade/la-metro-councilmatic/blob/b1cad35df52257e99235081d50e0c6f758e64cc4/lametro/templates/base.html#L102-L186
- The search view instantiates the search form with the incoming query dict, including the reCAPTCHA token.
https://github.com/datamade/la-metro-councilmatic/blob/b1cad35df52257e99235081d50e0c6f758e64cc4/lametro/views.py#L689-L690
...
https://github.com/datamade/la-metro-councilmatic/blob/b1cad35df52257e99235081d50e0c6f758e64cc4/lametro/views.py#L716
- The reCAPTCHA field on the form validates the token for us in the course of normal form validation, returning nothing if the token is invalid or the results otherwise.
https://github.com/datamade/la-metro-councilmatic/blob/b1cad35df52257e99235081d50e0c6f758e64cc4/lametro/forms.py#L68-L80

### Checklist


## Testing Instructions

 * Confirm CI passes.
 * This branch has been deployed to the staging site. Navigate around and confirm that everything – search or otherwise – behaves as normal: https://lametro-upgrade.datamade.us
